src/main/java/org/cryptomator/cryptofs/health/type/CiphertextFileTypeCheck.java

```diff
@@ -0,0 +1,160 @@
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * Checks for each c9r or c9s dir in the vault structure if its {@link org.cryptomator.cryptofs.common.CiphertextFileType} can be determined.
+ * <p>
+ * The type is based on the presence of a signature file.
+ * The signature files are {@value org.cryptomator.cryptofs.common.Constants#DIR_FILE_NAME}, {@value org.cryptomator.cryptofs.common.Constants#SYMLINK_FILE_NAME} and {@value org.cryptomator.cryptofs.common.Constants#CONTENTS_FILE_NAME}.
+ * Note: For c9r dirs, only the dir and symlink sig files are tested.
+ */
+public class CiphertextFileTypeCheck implements HealthCheck {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CiphertextFileTypeCheck.class);
+	private static final int MAX_TRAVERSAL_DEPTH = 3;
+
+	@Override
+	public String name() {
+		return "Resource Type Check";
+	}
+
+	@Override
+	public void check(Path pathToVault, VaultConfig config, Masterkey masterkey, Cryptor cryptor, Consumer<DiagnosticResult> resultCollector) {
+
+		// scan vault structure:
+		var dataDirPath = pathToVault.resolve(Constants.DATA_DIR_NAME);
+		var dirVisitor = new CiphertextFileTypeCheck.DirVisitor(resultCollector);
+		try {
+			Files.walkFileTree(dataDirPath, Set.of(), MAX_TRAVERSAL_DEPTH, dirVisitor);
+		} catch (IOException e) {
+			LOG.error("Traversal of data dir failed.", e);
+			resultCollector.accept(new CheckFailed("Traversal of data dir failed. See log for details."));
+			return;
+		}
+	}
+
+	// visible for testing
+	static class DirVisitor extends SimpleFileVisitor<Path> {
+
+		private final Consumer<DiagnosticResult> resultCollector;
+
+		public DirVisitor(Consumer<DiagnosticResult> resultCollector) {this.resultCollector = resultCollector;}
+
+		@Override
+		public FileVisitResult visitFile(Path dir, BasicFileAttributes attrs) {
+			switch (determineFileType(dir, attrs)) {
+				case C9R_DIR -> checkCiphertextTypeC9r(dir);
+				case C9S_DIR -> checkCiphertextTypeC9s(dir);
+				case UNRELATED -> {}
+			}
+			;
+			return FileVisitResult.CONTINUE;
+		}
+
+		DirType determineFileType(Path p, BasicFileAttributes attrs) {
+			var dirName = p.getFileName().toString();
+			boolean isDir = attrs.isDirectory();
+
+			if (isDir && dirName.endsWith(Constants.CRYPTOMATOR_FILE_SUFFIX)) {
+				return DirType.C9R_DIR;
+			} else if (isDir && dirName.endsWith(Constants.DEFLATED_FILE_SUFFIX)) {
+				return DirType.C9S_DIR;
+			} else {
+				return DirType.UNRELATED;
+			}
+		}
+
+		void checkCiphertextTypeC9r(Path dir) {
+			boolean isDir = containsDirFile(dir);
+			boolean isSymlink = containsSymlinkFile(dir);
+			var types = ciphertextFileTypesFrom(isDir, isSymlink, false);
+
+			if (isDir ^ isSymlink) {
+				resultCollector.accept(new KnownType(dir, types.iterator().next()));
+			} else if (isDir || isSymlink) {
+				resultCollector.accept(new AmbiguousType(dir, types));
+			} else {
+				resultCollector.accept(new UnknownType(dir));
+			}
```

Comment:
second branch is essentially `if not a xor b and a or b`, which can also be expressed as `if a and b` which is easier to read imho

Reply:
In fact, the easiest approach would probably be to just count the number of items in `types`.

Reply:
> just count the number of items in `types`:

Indeed a very good idea, thanks for the code snippet! Fixed in bd78457